### PR TITLE
fix memory leak in pilot

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes: #11180

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 1]
```diff
@@ -508,6 +508,18 @@ func (s *DiscoveryServer) edsUpdate(shard, serviceName string,
 	defer s.mutex.Unlock()
 	requireFull := false
 
+	// To prevent memory leak.
+	// Should delete the service EndpointShards, when endpoints deleted or service deleted.
+	if len(istioEndpoints) == 0 {
+		if s.EndpointShardsByService[serviceName] != nil {
+			delete(s.EndpointShardsByService[serviceName].Shards, shard)
```
Comment (YaoZengzeng) [inline pilot/pkg/proxy/envoy/v2/eds.go]: `s.EndpointShardsByService[serviceName].Lock()` is needed before deleting the shard.

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/eds.go]: good catch

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 2]
```diff
@@ -508,6 +510,21 @@ func (s *DiscoveryServer) edsUpdate(shard, serviceName string,
...
+	// To prevent memory leak.
+	// Should delete the service EndpointShards, when endpoints deleted or service deleted.
+	if len(istioEndpoints) == 0 {
+		if s.EndpointShardsByService[serviceName] != nil {
+			s.EndpointShardsByService[serviceName].mutex.Lock()
+			delete(s.EndpointShardsByService[serviceName].Shards, shard)
+			svcShards := len(s.EndpointShardsByService[serviceName].Shards)
+			s.EndpointShardsByService[serviceName].mutex.Unlock()
+			if svcShards == 0 {
+				delete(s.EndpointShardsByService, serviceName)
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/eds.go]: What do you mean?

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/eds.go]: s.mutex is to protect s.EndpointShardsByService

[the remark this answers is not in the document]
Comment (ymesika) [reply pilot/pkg/proxy/envoy/v2/eds.go]: Right, and in lines 205 and 163 you protect reading from here while here you modify it?

[the remark this answers is not in the document]
Comment (ymesika) [reply pilot/pkg/proxy/envoy/v2/eds.go]: Oh, you already locked it above..

**Comment by rshriram** [comment]
fix branch

**Comment by hzxuzhonghu** [comment]
This is definitely a bug fix. It will cause memory leak.

**Comment by hzxuzhonghu** [comment]
@wenchenglu Would you let it go into 1.1?

**Comment by wenchenglu** [comment]
@hzxuzhonghu 
yes, that is a critical fix, should go in to 1.1

Ping me once the required test passed

**Comment by hzxuzhonghu** [comment]
Weird, retest can not work

**Comment by hzxuzhonghu** [comment]
@wenchenglu It's ready now.

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 3]
```diff
@@ -534,6 +534,18 @@ func (s *DiscoveryServer) edsUpdate(shard, serviceName string,
 	defer s.mutex.Unlock()
 	requireFull := false
 
+	// To prevent memory leak.
+	// Should delete the service EndpointShards, when endpoints deleted or service deleted.
+	if len(istioEndpoints) == 0 {
+		if s.EndpointShardsByService[serviceName] != nil {
+			delete(s.EndpointShardsByService[serviceName].Shards, shard)
+			if len(s.EndpointShardsByService[serviceName].Shards) == 0 {
+				delete(s.EndpointShardsByService, serviceName)
```
Comment (JamazRuan) [inline pilot/pkg/proxy/envoy/v2/eds.go]: Now endPoint deletion not trigger ConfigUpdate?

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/eds.go]: We donot need this to trigger ConfigUpdate, because service deletion will do

Comment (JamazRuan) [reply pilot/pkg/proxy/envoy/v2/eds.go]: @hzxuzhonghu 
ConfigUpdate does not need to triggered when the endpoint change from 1 to 0 ?

**Pull request merged**